lib/crawler.js

```diff
@@ -230,42 +237,222 @@ util.inherits(Crawler, EventEmitter);
         });
-        crawler.running = true;
-        crawler.emit("crawlstart");
+    });
+
+    return crawler;
+};
+
+/*
+    Public: Generates a configuration object for http[s].request.
+
+    queueItem - Queue item representing resource to be fetched
+
+    Returns an object that can be passed directly to http[s].request.
+
+*/
+Crawler.prototype.getRequestOptions = function (queueItem) {
+    var crawler = this;
+
+    var agent = queueItem.protocol === "https" ? crawler.httpsAgent : crawler.httpAgent;
+
+    // Extract request options from queue;
+    var requestHost = queueItem.host,
+        requestPort = queueItem.port,
+        requestPath = queueItem.path;
+
+    // Are we passing through an HTTP proxy?
+    if (crawler.useProxy) {
+        requestHost = crawler.proxyHostname;
+        requestPort = crawler.proxyPort;
+        requestPath = queueItem.url;
+    }
+
+    // Load in request options
+    var requestOptions = {
+        method: "GET",
+        host: requestHost,
+        port: requestPort,
+        path: requestPath,
+        agent: agent,
+        headers: {
+            "User-Agent": crawler.userAgent,
+            "Host": queueItem.host + (queueItem.port !== 80 ? ":" + queueItem.port : "")
+        }
+    };
+
+    if (queueItem.referrer) {
+        requestOptions.headers.Referer = queueItem.referrer;
+    }
+
+    // If port is one of the HTTP/HTTPS defaults, delete the option to avoid conflicts
+    if (requestOptions.port === 80 || requestOptions.port === 443) {
+        delete requestOptions.port;
+    }
+
+    // Add cookie header from cookie jar if we're configured to
+    // send/accept cookies
+    if (crawler.acceptCookies && crawler.cookies.getAsHeader()) {
+        requestOptions.headers.cookie =
+            crawler.cookies.getAsHeader(queueItem.host, queueItem.path);
+    }
+
+    // Add auth headers if we need them
+    if (crawler.needsAuth) {
+        var auth = crawler.authUser + ":" + crawler.authPass;
+
+        // Generate auth header
+        auth = "Basic " + new Buffer(auth).toString("base64");
+        requestOptions.headers.Authorization = auth;
+    }
+
+    // Add proxy auth if we need it
+    if (crawler.proxyUser !== null && crawler.proxyPass !== null) {
+        var proxyAuth = crawler.proxyUser + ":" + crawler.proxyPass;
+
+        // Generate auth header
+        proxyAuth = "Basic " + new Buffer(proxyAuth).toString("base64");
+        requestOptions.headers["Proxy-Authorization"] = proxyAuth;
+    }
+
+    // And if we've got any custom headers available
+    if (crawler.customHeaders) {
+        for (var header in crawler.customHeaders) {
+            if (!crawler.customHeaders.hasOwnProperty(header)) {
+                continue;
+            }
+
+            requestOptions.headers[header] = crawler.customHeaders[header];
+        }
+    }
+
+    return requestOptions;
+};
+
+/*
+    Public: Initiates a request for the robots.txt resource on crawler's target
```

Comment:
I'm wondering whether we need a cached list of hostnames, with a parsed robots-txt for each one — if the crawler, using this patch, continues to another hostname, then it won't respect robots.txt over there.

(I'd consider merging this but adding that as a further issue to be addressed.)